### PR TITLE
LPS-176521 asset entry is not updated when the KB expires

### PR DESCRIPTION
**Pull request by AliciaGarciaGarcia**
https://issues.liferay.com/browse/LPS-176521

Instead of updating the kbArticle 2 times, or updating the 

The problem is the way we obtain the asset entry from the KBArticle. It depends on the approval of the previous kbArticle.
The KBArticle table stores all the versions of each article, but the asset only must point to the approved ones or the drafts. On the case when we expire the kbArticle, the status is not approved but the previous asset  still is, so we have to pickup this on to modify it.

- LPS-176521 update the asset entry and indexer if is expired too
- LPS-176521 remove old index for the old version of the KBArticle
- LPS-176521 extract the logic of the get action
- LPS-176521 not needed
- LPS-176521 use method to reindex
- LPS-176521 when we are expiring the KB Article without previous updating the asset entry, the classPK method return the not approved one, but the asset is not updated yet, so the classPK should be the one of the approved article
- LPS-176521 autoSF
- LPS-176521 add tests to check the behavior of the kb with the asset entry

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
ci:test:sf

**Comment by AliciaGarciaGarcia** [comment]
ci:test:knowledge-base

**Comment by AliciaGarciaGarcia** [comment]
ci:test:knowledge-base

**Comment by AliciaGarciaGarcia** [comment]
ci:test:relevant

**Review by adolfopa — APPROVED** [review]

**Comment by AliciaGarciaGarcia** [comment]
ci:reevaluate:1307401_5938

**Review by marcogalluzzi — CHANGES_REQUESTED** [review]
This change introduces a problem when you set an expiration date for a Draft article, it tries to get an Asset Entry when none exist -> It seems that when you save the first version of an article as a Draft, no Assent Entries are created.

This error occurs when the scheduled task tries to expire the Draft Article:

```
2023-03-07 16:49:34.644 ERROR [liferay/scheduler_dispatch-149][ParallelDestination:59] Unable to process message {destinationName=liferay/scheduler_dispatch, response=null, responseDestinationName=null, responseId=null, payload=null, values={GROUP_NAME=com.liferay.knowledge.base.internal.messaging.CheckKBArticleMessageListener, companyId=0, groupId=0, DESTINATION_NAME=liferay/scheduler_dispatch, JOB_STATE=com.liferay.portal.kernel.scheduler.JobState@7bfb9a1b, STORAGE_TYPE=MEMORY_CLUSTERED, JOB_NAME=com.liferay.knowledge.base.internal.messaging.CheckKBArticleMessageListener}}
com.liferay.portal.kernel.messaging.MessageListenerException: java.lang.NullPointerException
	at com.liferay.portal.kernel.messaging.BaseMessageListener.receive(BaseMessageListener.java:41) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.scheduler.messaging.SchedulerEventMessageListenerWrapper.receive(SchedulerEventMessageListenerWrapper.java:59) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.messaging.InvokerMessageListener.receive(InvokerMessageListener.java:74) ~[portal-kernel.jar:?]
	at com.liferay.portal.messaging.internal.ParallelDestination$1.run(ParallelDestination.java:56) [bundleFile:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_341]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_341]
	at java.lang.Thread.run(Thread.java:750) [?:1.8.0_341]
Caused by: java.lang.NullPointerException
	at com.liferay.knowledge.base.service.impl.KBArticleLocalServiceImpl.updateStatus(KBArticleLocalServiceImpl.java:1410) ~[?:?]
	at com.liferay.knowledge.base.service.impl.KBArticleLocalServiceImpl._expireKBArticlesByCompany(KBArticleLocalServiceImpl.java:1676) ~[?:?]
	at com.liferay.knowledge.base.service.impl.KBArticleLocalServiceImpl.lambda$_checkKBArticlesByExpirationDate$0(KBArticleLocalServiceImpl.java:1609) ~[?:?]
	at com.liferay.portal.service.impl.CompanyLocalServiceImpl.forEachCompany(CompanyLocalServiceImpl.java:500) ~[portal-impl.jar:?]
	at com.liferay.portal.service.impl.CompanyLocalServiceImpl.forEachCompany(CompanyLocalServiceImpl.java:478) ~[portal-impl.jar:?]
	at sun.reflect.GeneratedMethodAccessor726.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_341]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_341]
	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50) ~[portal-impl.jar:?]
	at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49) ~[portal-impl.jar:?]
	at com.sun.proxy.$Proxy87.forEachCompany(Unknown Source) ~[?:?]
	at com.liferay.knowledge.base.service.impl.KBArticleLocalServiceImpl._checkKBArticlesByExpirationDate(KBArticleLocalServiceImpl.java:1608) ~[?:?]
	at com.liferay.knowledge.base.service.impl.KBArticleLocalServiceImpl.checkKBArticles(KBArticleLocalServiceImpl.java:352) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_341]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_341]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_341]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_341]
	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50) ~[portal-impl.jar:?]
	at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69) ~[portal-impl.jar:?]
	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57) ~[portal-impl.jar:?]
	at com.liferay.change.tracking.internal.aop.CTTransactionAdvice.invoke(CTTransactionAdvice.java:79) ~[?:?]
	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57) ~[portal-impl.jar:?]
	at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49) ~[portal-impl.jar:?]
	at com.sun.proxy.$Proxy1394.checkKBArticles(Unknown Source) ~[?:?]
	at com.liferay.knowledge.base.internal.messaging.CheckKBArticleMessageListener.doReceive(CheckKBArticleMessageListener.java:75) ~[?:?]
	at com.liferay.portal.kernel.messaging.BaseMessageListener.receive(BaseMessageListener.java:31) ~[portal-kernel.jar:?]
	... 6 more
```

File: modules/apps/knowledge-base/knowledge-base-test/src/testIntegration/java/com/liferay/knowledge/base/service/test/KBArticleLocalServiceTest.java [hunk 1]
```diff
@@ -221,6 +223,57 @@ public void testAddDraftKBArticleInsideDraftKBArticle() throws Exception {
...
+
+		KBArticle kbArticle = _kbArticleLocalService.addKBArticle(
+			null, _user.getUserId(), _kbFolderClassNameId,
+			KBFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+			StringUtil.randomString(), StringUtil.randomString(),
+			StringUtil.randomString(), StringUtil.randomString(), null, null,
+			null, null, null, _serviceContext);
+
+		AssetEntry assetEntry = _assetEntryLocalService.getEntry(
+			KBArticle.class.getName(), kbArticle.getResourcePrimKey());
```
[the remark this answers is not in the document]
Comment (AliciaGarciaGarcia) [reply modules/apps/knowledge-base/knowledge-base-test/src/testIntegration/java/com/liferay/knowledge/base/service/test/KBArticleLocalServiceTest.java:238]: I do not think this is a good idea.

**Comment by AliciaGarciaGarcia** [comment]
rebase, made some changes on test names and additional commits after meeting

**Comment by AliciaGarciaGarcia** [comment]
ci:test:sf

**Comment by AliciaGarciaGarcia** [comment]
ci:test:knowledge-base

File: modules/apps/knowledge-base/knowledge-base-test/src/testIntegration/java/com/liferay/knowledge/base/service/test/KBArticleLocalServiceTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-test/src/testIntegration/java/com/liferay/knowledge/base/service/test/KBArticleLocalServiceTest.java]: Why not?

**Review by marcogalluzzi — APPROVED** [review]

**Comment by marcogalluzzi** [comment]
ci:reevaluate:1277401_7215

**Comment by AliciaGarciaGarcia** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
ci:forward

**Pull request closed without merging**